File: 2021/09/04/main.go
package main

func sumOfDistancesInTree(n int, edges [][]int) []int {
	tree := make([][]int, n)
	for _, e := range edges {
		i, j := e[0], e[1]
		tree[i] = append(tree[i], j)
		tree[j] = append(tree[j], i)
	}

	res := make([]int, n)
	count := make([]int, n)

	isSeen := make([]bool, n)
	deepFirstSearch(0, isSeen, count, res, tree)

	isSeen = make([]bool, n)
	deepFirstMove(0, isSeen, count, res, tree)

	return res
}

func deepFirstSearch(root int, isSeen []bool, count, res []int, tree [][]int) {
	isSeen[root] = true
	for _, n := range tree[root] {
		if isSeen[n] {
			continue
		}
		deepFirstSearch(n, isSeen, count, res, tree)
		count[root] += count[n]
		res[root] += res[n] + count[n]
	}
	count[root]++
}

func deepFirstMove(root int, isSeen []bool, count, res []int, tree [][]int) {
	N := len(isSeen)
	isSeen[root] = true
	for _, n := range tree[root] {
		if isSeen[n] {
			continue
		}
		res[n] = res[root] + (N - count[n]) - count[n]
		deepFirstMove(n, isSeen, count, res, tree)
	}
}

File: 2021/09/04/main.py
class Solution:
    def sumOfDistancesInTree(self, n: int, edges: List[List[int]]) -> List[int]:
        tree = collections.defaultdict(set)
        res = [0] * n
        count = [0] * n
        for i, j in edges:
            tree[i].add(j)
            tree[j].add(i)

        def dfs(root=0, seen=set()):
            seen.add(root)
            for i in tree[root]:
                if i not in seen:
                    dfs(i, seen)
                    count[root] += count[i]
                    res[root] += res[i] + count[i]
            count[root] += 1

        def dfm(root=0, seen=set()):
            seen.add(root)
            for i in tree[root]:
                if i not in seen:
                    res[i] = res[root]+ n - 2 * count[i]
                    dfm(i, seen)
        dfs()
        dfm()
        return res